src/openrct2/actions/RideCreateAction.hpp

```diff
@@ -0,0 +1,314 @@
+{
+public:
+    sint32 rideType;
+    sint32 rideSubType;
+
+    void Serialise(DataSerialiser& stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << rideType << rideSubType;
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if (rideType >= RIDE_TYPE_COUNT)
+        {
+            return std::make_unique<RideCreateGameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_INVALID_RIDE_TYPE);
+        }
+
+        return std::make_unique<RideCreateGameActionResult>();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_ride_entry * rideEntry;
+        auto res = std::make_unique<RideCreateGameActionResult>();
+
+        sint32 subType = rideSubType;
+        if (subType == RIDE_ENTRY_INDEX_NULL)
+        {
+            uint8 *availableRideEntries = get_ride_entry_indices_for_ride_type(rideType);
+            for (uint8 *rei = availableRideEntries; *rei != RIDE_ENTRY_INDEX_NULL; rei++)
+            {
+                rideEntry = get_ride_entry(*rei);
+
+                // Can happen in select-by-track-type mode
+                if (!ride_entry_is_invented(*rei) && !gCheatsIgnoreResearchStatus)
+                {
+                    continue;
+                }
+
+                if (!(rideEntry->flags & RIDE_ENTRY_FLAG_SEPARATE_RIDE_NAME) || rideTypeShouldLoseSeparateFlag(rideEntry))
+                {
+                    subType = *rei;
+                    break;
+                }
+            }
+            if (subType == RIDE_ENTRY_INDEX_NULL)
+            {
+                subType = availableRideEntries[0];
+                if (subType == RIDE_ENTRY_INDEX_NULL)
+                {
+                    res->Error = GA_ERROR::INVALID_PARAMETERS;
+                    res->ErrorMessage = STR_INVALID_RIDE_TYPE;
+                    return std::move(res);
+                }
+            }
+        }
+
+        if (subType >= 128)
+        {
+            res->Error = GA_ERROR::INVALID_PARAMETERS;
+            res->ErrorMessage = STR_INVALID_RIDE_TYPE;
+            return std::move(res);
+        }
+
+        sint32 rideEntryIndex = subType;
+        sint32 rideIndex = ride_get_empty_slot();
+        if (rideIndex == -1) 
+        {
+            res->Error = GA_ERROR::DISALLOWED;
+            res->ErrorMessage = STR_TOO_MANY_RIDES;
+            return std::move(res);
+        }
+
+        res->rideIndex = rideIndex;
+        res->rideColor = ride_get_random_colour_preset_index(rideType) | (ride_get_unused_preset_vehicle_colour(rideType, subType) << 8);
+
+        auto ride = get_ride(rideIndex);
+        rideEntry = get_ride_entry(rideEntryIndex);
+        if (rideEntry == nullptr || rideEntry == (rct_ride_entry *)-1)
+        {
+            log_warning("Invalid request for ride %u", rideIndex);
+            res->Error = GA_ERROR::UNKNOWN;
+            res->ErrorMessage = STR_UNKNOWN_OBJECT_TYPE;
+            return std::move(res);
```

Comment:
But in Query we have no rideIndex yet which is first obtained by ride_get_empty_slot()